stingray/stats.py

```diff
@@ -0,0 +1,268 @@
+
+
+__all__ = ['p_multitrial_from_single_trial',
+           'p_single_trial_from_p_multitrial',
+           'fold_profile_probability',
+           'fold_detection_level',
+           'z2_n_detection_level',
+           'z2_n_probability']
+
+
+def p_multitrial_from_single_trial(p1, n):
+    """Calculate a multi-trial probability from a single-trial one.
+
+    Calling _p_ the probability of a single hit, the Binomial
+    distributions says that the probability _at least_ one outcome
+    in n trials is
+
+                         n
+    P (k ≥ 1) =   Σ    (   ) p^k (1 - p)^(n - k)
+                k ≥ 1    k
+
+    or more simply, using P(k ≥ 0) = 1
+                          n
+    P (k ≥ 1) =   1 -   (   ) (1 - p)^n = 1 - (1 - p)^n
+                          0
+    """
+    return 1 - (1 - np.longdouble(p1))**np.longdouble(n)
+
+
+def p_single_trial_from_p_multitrial(pn, n):
+    """Calculate a single-trial probability from a multi-trial one.
+
+    How to calculate a single-trial probability when we only have
+    a multi-trial one? This might be the case because, e.g., we
+    want to have a 10% probability of detecting a signal in an
+    entire power spectrum, and we need to correct the detection
+    level accordingly.
+
+    The typical procedure is dividing the initial probability
+    (often called _epsilon_) by the number of trials. This is
+    often a good approximation, but it is not formally correct.
+    If epsilon * n approaches 1, this approximation gives
+    incorrect results.
+
+    Here we calculate this probability by inverting the Binomial
+    problem. Given that (see ``p_multitrial_from_single_trial``)
+    the probability of getting more than one hit in n trials,
+    given the single-trial probability _p_, is
+
+    P (k ≥ 1) =  1 - (1 - p)^n,
+
+    we get the single trial probability from the multi-trial one
+    from
+
+    p = 1 - (1 - P)^1/n
+
+    """
+    if isinstance(n, Iterable):
+        return np.array([p_single_trial_from_p_multitrial(pn, ni)
+                         for ni in n])
+
+    # Numerical errors arise when pn is very close to 1.
+    if 1 - pn < np.finfo(np.longdouble).resolution * 1000:
+        warnings.warn("Multi-trial probability is very close to 1.")
+        warnings.warn("The problem is ill-conditioned. Returning NaN")
+        return np.nan
+
+    p1 = 1 - np.power(1 - np.longdouble(pn), 1/np.longdouble(n))
+    return p1
+
+
+def fold_profile_probability(stat, nbin, ntrial=1):
+    """Calculate the probability of a certain folded profile, due to noise.
+
+    Parameters
+    ----------
+    stat : float
+        The epoch folding statistics
+    nbin : int
+        The number of bins in the profile
+
+    Other Parameters
+    ----------------
+    ntrial : int
+        The number of trials executed to find this profile
+
+    Returns
+    -------
+    p : float
+        The probability that the profile has been produced by noise
+    """
+    p1 = stats.chi2.sf(stat, (nbin - 1))
```

Comment:
Should this really be the Chi-square distribution or the distributions derived in Groth (1975)?

Reply:
They are exactly equivalent in the absence of source power, as far as I understand.